src/data/set/Union_lift.lean

```diff
@@ -0,0 +1,170 @@
+/-
+Copyright (c) 2021 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+import data.set.lattice
+import order.directed
+/-!
+# Union lift
+This file defines `set.Union_lift` to glue together functions defined on each of a collection of
+sets to make a function on the Union of those sets.
+
+## Main definitions
+
+* `set.Union_lift` -  Given a Union of sets `Union S`, define a function on any subset of the Union
+  by defining it on each component, and proving that it agrees on the intersections.
+* `set.lift_cover` - Version of `set.Union_lift` for the special case that the sets cover the
+  entire type.
+
+## Main statements
+
+There are proofs of the obvious properties of `Union_lift`, i.e. what it does to elements of
+each of the sets in the `Union`, stated in different ways.
+
+There are also three lemmas about `Union_lift` intended to aid with proving that `Union_lift` is a
+homomorphism when defined on a Union of substructures. There is one lemma each to show that
+constants, unary functions, or binary functions are preserved. These lemmas are:
+
+*`set.Union_lift_const`
+*`set.Union_lift_unary`
+*`set.Union_lift_binary`
+
+## Tags
+
+directed union, directed supremum, glue, gluing
+-/
+
+variables {α ι β : Type*}
+
+namespace set
+
+section Union_lift
+
+/- The unused argument `hf` is left in the definition so that the `simp` lemmas
+`Union_lift_inclusion` will work without the user having to provide `hf` explicitly to
+simplify terms involving `Union_lift`. -/
+/-- Given a Union of sets `Union S`, define a function on the Union by defining
+it on each component, and proving that it agrees on the intersections. -/
+@[nolint unused_arguments]
+noncomputable def Union_lift (S : ι → set α)
+  (f : Π i (x : S i), β)
+  (hf : ∀ i j (x : α) (hxi : x ∈ S i) (hxj : x ∈ S j), f i ⟨x, hxi⟩ = f j ⟨x, hxj⟩)
+  (T : set α) (hT : T ⊆ Union S) (x : T) : β :=
+let i := classical.indefinite_description _ (mem_Union.1 (hT x.prop)) in
+f i ⟨x, i.prop⟩
+
+variables
+  {S : ι → set α}
+  {f : Π i (x : S i), β}
+  {hf : ∀ i j (x : α) (hxi : x ∈ S i) (hxj : x ∈ S j), f i ⟨x, hxi⟩ = f j ⟨x, hxj⟩}
+  {T : set α} {hT : T ⊆ Union S} (hT' : T = Union S)
```

Comment:
In some sense, I guess you could call `hT'` a "characteristic predicate" on for `Union S`. It probably doesn't make sense to generalize this to an equiv of types, does it?

Reply:
I don't think so. An equality of sets is a bit stronger than an `equiv` in the sense that it comes with a bunch of inclusion maps out of the `S i` that commute with the `equiv`. If it took an `equiv` as an argument, all of the `simp` lemmas about it would have to contain the `equiv` on the rhs I think.